```vast-cc``` met error: redefinition of symbol named 'fscanf'

Thank you very much for making changes to the code for ```csmith```. I used the latest ```vast-cc``` and executed ```vast-cc --ccopts -xc --from-source  r1.c``` . (```r1.c``` contains the header file of ```csmith```).

**r1.c:**
```console
#define NO_LONGLONG

#include "csmith.h"

static long __undefined;

static uint8_t  func_1(void);

static uint8_t  func_1(void)
{ 
    int32_t l_4 = 0;
    int32_t l_5[7] = {0,0,0,0,0,0,0};
    int i;
    l_5[6] = (safe_add_func_uint32_t_u_u(4294967293U, l_4));
    return l_5[6];
}

int main (int argc, char* argv[])
{
    int print_hash_value = 0;
    if (argc == 2 && strcmp(argv[1], "1") == 0) print_hash_value = 1;
    platform_main_begin();
    crc32_gentab();
    func_1();
    platform_main_end(crc32_context ^ 0xFFFFFFFFUL, print_hash_value);
    return 0;
}
```
**And it returned with :**
```console
In file included from input.cc:4:
In file included from ./csmith.h:45:
In file included from ./random_inc.h:100:
./safe_math.h:1201:36: warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
    ((sf1 <= INT32_MIN) || (sf1 >= INT32_MAX)) ?
                                ~~ ^~~~~~~~~
/usr/include/stdint.h:123:22: note: expanded from macro 'INT32_MAX'
# define INT32_MAX              (2147483647)
                                 ^~~~~~~~~~
unknown:1:1: error: redefinition of symbol named 'fscanf'
unknown:1:1: note: see current operation: "hl.func"() ({
}) {function_type = (!hl.lvalue<!hl.ptr<!hl.typedef<"FILE">,  restrict >>, !hl.lvalue<!hl.ptr<!hl.char< const >,  restrict >>) -> !hl.int, linkage = 0 : i32, sym_name = "fscanf", sym_visibility = "private"} : () -> ()
/usr/include/stdio.h:415:12: note: see existing symbol definition here
extern int fscanf (FILE *__restrict __stream,
```
However, I can compile r1.c with gcc correctly. Could anyone know what should I do next?

--- a/include/vast/Dialect/HighLevel/HighLevelOps.td
+++ b/include/vast/Dialect/HighLevel/HighLevelOps.td
@@ -361,15 +361,18 @@ def ReturnOp
   ];
 }
 
+// use InferTypeOpInterface
 def DeclRefOp
-  : HighLevel_Op< "ref", [NoSideEffect, SameOperandsAndResultType] >
+  : HighLevel_Op< "ref", [NoSideEffect,
+      DeclareOpInterfaceMethods<InferTypeOpInterface>
+  ] >
   , Arguments<(ins AnyType:$decl)>
-  , Results<(outs AnyType:$result)>
+  , Results<(outs LValueOf<AnyType>:$result)>
 {
   let summary = "VAST variable reference declaration";
   let description = [{ VAST variable reference declaration }];
 
-  let assemblyFormat = "$decl attr-dict `:` type($decl)";
+  let assemblyFormat = "$decl attr-dict `:`type(operands)";
 }
 
 def FuncRefOp

--- a/include/vast/Translation/CodeGen.hpp
+++ b/include/vast/Translation/CodeGen.hpp
@@ -101,6 +101,7 @@ namespace vast::cg
         using TypeDeclsScope     = ScopedSymbolTable< const clang::TypeDecl *, hl::TypeDeclOp >;
         using EnumDeclsScope     = ScopedSymbolTable< const clang::EnumDecl *, hl::EnumDeclOp >;
         using EnumConstantsScope = ScopedSymbolTable< const clang::EnumConstantDecl *, hl::EnumConstantOp >;
+        using LabelTable         = ScopedSymbolTable< const clang::LabelDecl*, hl::LabelDeclOp >;
         using FunctionsScope     = ScopedSymbolTable< const clang::FunctionDecl *, hl::FuncOp >;
         using VariablesScope     = ScopedSymbolTable< const clang::VarDecl *, Value >;
 
@@ -109,6 +110,7 @@ namespace vast::cg
             TypeDeclsScope     typedecls;
             EnumDeclsScope     enumdecls;
             EnumConstantsScope enumconsts;
+            LabelTable         labels;
             FunctionsScope     funcdecls;
             VariablesScope     globs;
         };
@@ -604,6 +606,7 @@ namespace vast::cg
                 .typedecls  = _cgctx->typedecls,
                 .enumdecls  = _cgctx->enumdecls,
                 .enumconsts = _cgctx->enumconsts,
+                .labels     = _cgctx->labels,
                 .funcdecls  = _cgctx->funcdecls,
                 .globs      = _cgctx->vars
             });

--- a/include/vast/Translation/CodeGenContext.hpp
+++ b/include/vast/Translation/CodeGenContext.hpp
@@ -180,7 +180,7 @@ namespace vast::cg
         }
 
         hl::FuncOp lookup_function(const clang::FunctionDecl *decl, bool with_error = true) {
-            return symbol(funcdecls, decl, "error: undeclared function '" + decl->getName() + "'", with_error);
+            return symbol(funcdecls, decl, "undeclared function '" + decl->getName() + "'", with_error);
         }
 
         hl::FuncOp declare(const clang::FunctionDecl *decl, auto vast_decl_builder) {

--- a/include/vast/Translation/CodeGenDeclVisitor.hpp
+++ b/include/vast/Translation/CodeGenDeclVisitor.hpp
@@ -92,7 +92,10 @@ namespace vast::cg {
 
             // make function header, that will be later filled with function body
             // or returned as declaration in the case of external function
-            auto fn = make< hl::FuncOp >(loc, mangled_name.name, fty, linkage);
+            auto fn = context().declare(function_decl, [&] () {
+                return make< hl::FuncOp >(loc, mangled_name.name, fty, linkage);
+            });
+
             VAST_CHECK(fn.isDeclaration(), "expected empty body");
 
             mlir::SymbolTable::setSymbolVisibility(
@@ -150,8 +153,6 @@ namespace vast::cg {
                     }
                 }
 
-                type.dump();
-                fn.getFunctionType().dump();
                 if (fn && fn.getFunctionType() == type) {
                     return fn;
                 }
@@ -295,6 +296,10 @@ namespace vast::cg {
 
         // FIXME: remove as this duplicates logic from codegen driver
         operation VisitFunctionDecl(const clang::FunctionDecl *decl) {
+            if (auto fn = context().lookup_function(decl, false /* emit no error */)) {
+                return fn;
+            }
+
             InsertionGuard guard(builder());
             auto is_definition = decl->doesThisDeclarationHaveABody();
 

--- a/lib/vast/Dialect/HighLevel/HighLevelOps.cpp
+++ b/lib/vast/Dialect/HighLevel/HighLevelOps.cpp
@@ -154,6 +154,26 @@ namespace vast::hl
         }
     }
 
+    logical_result DeclRefOp::inferReturnTypes(
+        mcontext_t* mctx,
+        llvm::Optional< Location > /* loc */,
+        mlir::ValueRange operands,
+        mlir::DictionaryAttr attributes,
+        mlir::RegionRange /* regs */,
+        llvm::SmallVectorImpl< mlir_type > &rty
+    ) {
+        DeclRefOp::Adaptor op(operands, attributes);
+        auto type = op.getDecl().getType();
+
+        if (auto ltype = type.dyn_cast< LValueType >()) {
+            rty.push_back(ltype);
+        } else {
+            rty.push_back(LValueType::get(mctx, type));
+        }
+
+        return mlir::success();
+    }
+
     FoldResult ConstantOp::fold(mlir::ArrayRef<Attribute> operands) {
         VAST_CHECK(operands.empty(), "constant has no operands");
         return getValue();

